### DISCUSSION TITLE
should I expect a package to include its link dependencies

### DISCUSSION THREAD

**Discussion by willie** [microsoft/vcpkg#36934] [Q&A] [answered]
For example, `opentelemetry-cpp` package has a dependency on `protobuf`, but it fails at link time because it cannot find `libprotobuf`. If I add protobuf[^1]

```
find_package(Protobuf CONFIG REQUIRED)
target_link_libraries(${CMAKE_PROJECT_NAME} PRIVATE protobuf::libprotobuf)
```
to my project's `CMakeLists.txt` file, then it links correctly. Should I expect the package to already know this dependency? Does this vary from package to package? If I were to try and fix this for the packages I encounter this with, what's the correct way to start?

[^1]: protobuf throws a bunch of errors if you use lowercase `find_package(protobuf CONFIG REQUIRED)`, which is another issue, like how do you fix this without breaking existing usage?

**Comment by Osyotr** [comment, marked as answer]
It's upstream responsibility to add required `find_dependency` calls into their exported configs. Some packages do it properly, some don't. Vcpkg does try to fix such problems (though I don't think it's explicitly mentioned anywhere). `opentelementry-cpp` already has one [patch](https://github.com/microsoft/vcpkg/blob/a51a6c09e59eba72366c40344e07f9972f1b442b/ports/opentelemetry-cpp/add-missing-find-dependency.patch) that adds required `find_dependency` call, though not for `protobuf`.
The fix varies from port to port, check out github search for `find_dependency` to gather some inspiration: [link](https://github.com/search?q=repo%3Amicrosoft%2Fvcpkg%20find_dependency&type=code).

> protobuf throws a bunch of errors if you use lowercase find_package(protobuf CONFIG REQUIRED), which is another issue, like how do you fix this without breaking existing usage?

What errors? CMake searches for a file called `<lowercasePackageName>-config.cmake` or `<PackageName>Config.cmake`, ~~thus `find_package(protobuf CONFIG REQUIRED)` is the correct usage, because installed config file name is `protobuf-config.cmake`.~~ see https://github.com/microsoft/vcpkg/discussions/36934#discussioncomment-8601739

**Reply by willie** [reply]
Ok, it's not a bunch of errors, but if you do:

```
find_package(protobuf CONFIG REQUIRED)
```

You get this warning.

```
[cmake] CMake Warning (dev) at /opt/homebrew/Cellar/cmake/3.28.3/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:438 (message):
[cmake]   The package name passed to `find_package_handle_standard_args` (Protobuf)
[cmake]   does not match the name of the calling package (protobuf).  This can lead
[cmake]   to problems in calling code that expects `find_package` result variables
[cmake]   (e.g., `_FOUND`) to follow a certain pattern.
[cmake] Call Stack (most recent call first):
[cmake]   build/vcpkg_installed/arm64-osx/share/protobuf/protobuf-module.cmake:162 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
[cmake]   build/vcpkg_installed/arm64-osx/share/protobuf/protobuf-config.cmake:168 (include)
[cmake]   build/vcpkg_installed/arm64-osx/share/protobuf/vcpkg-cmake-wrapper.cmake:16 (_find_package)
[cmake]   vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
[cmake]   CMakeLists.txt:36 (find_package)
[cmake] This warning is for project developers.  Use -Wno-dev to suppress it.
```

You don't get this warning if you do this:

```
find_package(Protobuf CONFIG REQUIRED)
```

**Reply by Osyotr** [reply]
Well, protobuf does some questionable cmake shenanigans. It installs `protobuf-config.cmake` so you must use `find_package(protobuf)`. But then it [forwards](https://github.com/protocolbuffers/protobuf/blob/ab127f75015c595bf27dd75f8f22a64f4fd9c7c0/cmake/protobuf-config.cmake.in#L15-L18) you to something that looks like a find module that tells that it [found](https://github.com/protocolbuffers/protobuf/blob/ab127f75015c595bf27dd75f8f22a64f4fd9c7c0/cmake/protobuf-module.cmake.in#L161-L165) Protobuf.
I.e. you asked cmake to find `protobuf` (because you must use that name due to cmake rules) but it reported that it found `Protobuf`.
Try setting `protobuf_MODULE_COMPATIBLE` to false before invoking `find_package(protobuf)`.

Related:
https://github.com/protocolbuffers/protobuf/issues/8008#issuecomment-832595729

**Reply by willie** [reply]
I'll try that, but, then again, passing `Protobuf` works fine, no errors, so I'm tempted to just let it stay that way.

I wish that OpenTelemetry just handled this dependency itself.

**Reply by Osyotr** [reply]
> no errors

Unless you try to build it on linux

**Reply by willie** [reply]
Actually, yes, exact same behavior on Linux.

```
#17 1.559 -- Found Protobuf: /saturnv/build/vcpkg_installed/arm64-linux/tools/protobuf/protoc (found version "3.21.12.0") 
#17 1.559 CMake Warning (dev) at /usr/share/cmake-3.28/Modules/FindPackageHandleStandardArgs.cmake:438 (message):
#17 1.559   The package name passed to `find_package_handle_standard_args` (Protobuf)
#17 1.559   does not match the name of the calling package (protobuf).  This can lead
#17 1.559   to problems in calling code that expects `find_package` result variables
#17 1.559   (e.g., `_FOUND`) to follow a certain pattern.
#17 1.559 Call Stack (most recent call first):
#17 1.559   build/vcpkg_installed/arm64-linux/share/protobuf/protobuf-module.cmake:162 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
#17 1.559   build/vcpkg_installed/arm64-linux/share/protobuf/protobuf-config.cmake:168 (include)
#17 1.559   build/vcpkg_installed/arm64-linux/share/protobuf/vcpkg-cmake-wrapper.cmake:16 (_find_package)
#17 1.559   vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
#17 1.559   CMakeLists.txt:357 (find_package)
#17 1.559 This warning is for project developers.  Use -Wno-dev to suppress it.
```

**Reply by Osyotr** [reply]
Does `find_package(Protobuf CONFIG)` works for you on Linux?

**Reply by dg0yt** [reply]
From what I observe here, my conclusion is that the official usage must be
~~~cmake
find_package(Protobuf [CONFIG])
~~~
which is a supported package name for both the config and the find module.
However, because upstream chose the case-insensitive spelling for the cmake config filename, the vcpkg tool usage heuristics is unable to suggest the desired spelling.

**Reply by willie** [reply]
> Does `find_package(Protobuf CONFIG)` works for you on Linux?

Yes.

**Comment by mail96350** [comment]


**Reply by willie** [reply]
Please delete this.

**Comment by mail96350** [comment]
You mean my comment ? Why?

**Reply by willie** [reply]
Because it's irrelevant to the topic of this discussion.

**Comment by mail96350** [comment]
Ok I will delete it but tell me how to create a new topic to post this request

**Comment by StarGate-One** [comment]
@mail96350 

>Ok I will delete it but tell me how to create a new topic to post this request

See #30604 How to report a build failure [explanation video]

[Start here](https://github.com/microsoft/vcpkg/issues/new/choose)

**Comment by willie** [comment]
> It's upstream responsibility to add required `find_dependency` calls into their exported configs. Some packages do it properly, some don't. Vcpkg does try to fix such problems (though I don't think it's explicitly mentioned anywhere). `opentelementry-cpp` already has one [patch](https://github.com/microsoft/vcpkg/blob/a51a6c09e59eba72366c40344e07f9972f1b442b/ports/opentelemetry-cpp/add-missing-find-dependency.patch) that adds required `find_dependency` call, though not for `protobuf`. The fix varies from port to port, check out github search for `find_dependency` to gather some inspiration: [link](https://github.com/search?q=repo%3Amicrosoft%2Fvcpkg%20find_dependency&type=code).

Thanks for this answer, BTW. It was the real answer to the question, but given the direction we took about `Protobuf`, I marked that one. I will try to figure out the `find_dependency` usage. I appreciate the guidance.

**Reply by dg0yt** [reply]
Please mark the real answer. My point was that port protobuf needs an explicit usage file, to end confusion.

**Comment by willie** [comment]
Also related, people noting the same issue on the upstream repo (just recently, though): https://github.com/open-telemetry/opentelemetry-cpp/issues/2528
